### PR TITLE
Add script to expressvpn uninstall stanza

### PR DESCRIPTION
**Pull request by ruiokada**
ExpressVPN seems to have their own uninstall script [\[1\]](https://www.expressvpn.com/support/vpn-setup/app-for-mac-os-x/#uninstall) so I've updated the uninstall stanza to use this. I've attached to script to the latest version below. In addition to uninstalling the app, the script uninstalls the agent and update LaunchAgents (homebrew-cask's stanza also did previously so I've removed this part) and zaps a whole bunch of files (homebrew-cask's stanza didn't do previously). I have left the pkgutil part of the stanza, I am unsure if the way the script removes the app is equivalent to homebrew-cask's pkgutil.

After making all changes to a cask, verify:

- [x] `brew cask audit --download {{cask_file}}` is error-free.
- [x] `brew cask style --fix {{cask_file}}` reports no offenses.
- [x] There are no [open pull requests](https://github.com/Homebrew/homebrew-cask/pulls) for the same update.
- [x] The submission is for [a stable version](https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#stable-versions) or [documented exception](https://github.com/Homebrew/homebrew-cask/blob/master/doc/development/adding_a_cask.md#but-there-is-no-stable-version).

```
#!/usr/bin/env bash

#----------------------------------------------------------------------
# Functions
#----------------------------------------------------------------------
# Exits the script with the given exit code after waiting
# for a keypress.
#
# @param [Integer] $1 exit code.
function key_exit() {
    echo "Press any key to exit."
    read
    exit $1
}

# Appends a value to an array.
#
# @param [String] $1 Name of the variable to modify
# @param [String] $2 Value to append
function append() {
    eval $1[\${#$1[*]}]=$2
}

#----------------------------------------------------------------------
# Script
#----------------------------------------------------------------------
# Collect the directories and files to remove
system_files=()
user_files=()
append system_files "/Applications/ExpressVPN.app"

# 4.x files
append system_files "$HOME/Library/Application\ Support/com.expressvpn.ExpressVPN"

# 3.x files
append user_files "$HOME/Library/Application\ Support/ExpressVPN"

# staging files
append system_files "$HOME/Library/Application\ Support/com.expressvpn.ExpressVPN-staging"

# preferences
append user_files "$HOME/Library/Preferences/com.expressvpn.ExpressVPN.plist"
append user_files "$HOME/Library/Preferences/com.expressvpn.ExpressVPNGroup.plist"
append user_files "$HOME/Library/Preferences/group.com.expressvpn.ExpressVPN.plist"

# logs
append user_files "$HOME/Library/Logs/ExpressVPN"
append system_files "/EV.script.log"
append system_files "/var/log/EV.script.log"
append system_files "/var/log/xvpn.postinstall.log"

# kexts
append system_files "/Library/Extensions/tun.kext"
append system_files "/Library/Extensions/tap.kext"
append system_files "/Library/Extensions/ExpressVPNSplitTunnel.kext"

# shortcuts
append user_files "$HOME/Documents/ExpressVPN\ Shortcuts"

# chrome extensions
append user_files "$HOME/Library/Application\ Support/Google/Chrome/NativeMessagingHosts/com.expressvpn.helper.json"

# firefox extension
append user_files "$HOME/Library/Application\ Support/Mozilla/NativeMessagingHosts/com.expressvpn.helper.json"

# config files
append user_files "$HOME/.expressvpn.conf"

# launch agent
append system_files "$HOME/Library/LaunchAgents/com.expressvpn.ExpressVPN.agent.plist"
append system_files "$HOME/Library/LaunchAgents/com.expressvpn.ExpressVPN.update.plist"

my_files=("${system_files[@]}" "${user_files[@]}")

# Print the files and directories that are to be removed and verify
# with the user that that is what he/she really wants to do.
echo "The following files and directories will be removed:"
for file in "${my_files[@]}"; do
    if [ -e "$file" ]; then
        echo "    $file"
    fi
done

echo ""
echo -n "Do you wish to uninstall ExpressVPN (Yes/No)? "
read my_answer
my_answer=$(echo "$my_answer" | awk '{print tolower($0)}')
if [ "$my_answer" != "yes" ]; then
    echo "Aborting install. (answer: ${my_answer})"
    key_exit 2
fi

# Unload updater, note that we should always quit the updater first since
# it might interrupt the uninstallation process if it tries to install update
# when ExpressVPN.app is quiting
if [ -f "$HOME/Library/LaunchAgents/com.expressvpn.ExpressVPN.update.plist" ]; then
    /bin/launchctl unload -w "$HOME/Library/LaunchAgents/com.expressvpn.ExpressVPN.update.plist" > /dev/null 2>&1

    # wait until updater process exit
    while /usr/bin/pgrep -x "ExpressVPN Update" > /dev/null; do sleep 1; done
fi

# Quit running app
/usr/bin/osascript -e 'quit app "ExpressVPN"'

# Remove ikev2 configurations
/Applications/ExpressVPN.app/Contents/MacOS/IKEHelper.app/Contents/MacOS/expressvpn-ikehelper --remove > /dev/null 2>&1

# Initiate the actual uninstall, which requires admin privileges.
echo
echo "The uninstallation process requires administrative privileges"
echo "because some of the installed files cannot be removed by a"
echo "normal user. You may now be prompted for a password..."
echo

# Use AppleScript so we can use a graphical `sudo` prompt.
# This way, people can enter the username they wish to use
# for sudo, and it is more Apple-like.
tmpfile="/Applications/ExpressVPN.app/Contents/Resources/xvpn_uninstall.sh"

# Unload engine agent
if [ -f "$HOME/Library/LaunchAgents/com.expressvpn.ExpressVPN.agent.plist" ]; then
    /bin/launchctl unload -w "$HOME/Library/LaunchAgents/com.expressvpn.ExpressVPN.agent.plist" > /dev/null 2>&1

    # wait until engine process exit
    while /usr/bin/pgrep -x expressvpnd > /dev/null; do sleep 1; done
fi

/usr/bin/osascript -e "do shell script \"/bin/bash $tmpfile\" with administrator privileges"

# following commands should be done as normal user.

# remvoe all user files
for file in "${user_files[@]}"; do
if [ -e "$file" ]; then
    /bin/rm -rf "$file"
fi
done

# remove all preferences
defaults delete com.expressvpn.ExpressVPN
defaults delete com.expressvpn.ExpressVPN-Update

# delete previous client certificates
/usr/bin/security delete-certificate -Z 'BDE98D55300F49FC535A37286FA8871924971683' > /dev/null 2>&1

/usr/bin/security delete-certificate -Z '0F846F6BA98D2A5ADFECADD029C55BB589DB0301' > /dev/null 2>&1

/usr/bin/security delete-identity -c 'ExpressVPN Client' > /dev/null 2>&1

/usr/bin/security delete-certificate -c 'ExpressVPN Client' > /dev/null 2>&1

# Kill all Safari companion processes
/usr/bin/pkill -9 -f 'ExpressVPN mini' > /dev/null 2>&1

# Verify that the uninstall succeeded by checking whether every file
# we meant to remove is actually removed.
for file in "${my_files[@]}"; do
    if [ -e "${file}" ]; then
        echo "An error must have occurred since a file that was supposed to be"
        echo "removed still exists: ${file}"
        echo ""
        echo "Please try again."
        key_exit 1
    fi
done

echo "Successfully uninstalled ExpressVPN."
echo "Done."

# cleanup
if [ -e $tmpfile ]; then
    rm -f $tmpfile
fi

exit 0
```

### CODE REVIEW

File: Casks/expressvpn.rb [hunk 1]
```diff
@@ -11,9 +11,9 @@
 
   pkg "expressvpn_mac_#{version}_release.pkg"
 
-  uninstall pkgutil:   'com.expressvpn.ExpressVPN',
-            launchctl: [
-                         'com.expressvpn.ExpressVPN.agent',
-                         'com.expressvpn.ExpressVPN.update',
-                       ]
+  uninstall script:  {
+                       executable: '/Applications/ExpressVPN.app/Contents/Resources/uninstall.tool',
```
Comment (miccal) [inline Casks/expressvpn.rb:15]: Please change this to:
```
executable: "#{appdir}/ExpressVPN.app/Contents/Resources/uninstall.tool",
```

File: Casks/expressvpn.rb [hunk 2]
```diff
@@ -11,9 +11,9 @@
...
+                       input:      ['Yes'],
```
Comment (miccal) [inline Casks/expressvpn.rb:16]: Please add:
```
sudo:           true,
```

File: Casks/expressvpn.rb (hunk 1 above)
Comment (ruiokada) [reply Casks/expressvpn.rb]: Got it!

File: Casks/expressvpn.rb (hunk 2 above)
Comment (ruiokada) [reply Casks/expressvpn.rb:16]: Updated in latest commit.

**Comment by ruiokada** [comment]
Despite the sudo privileges, the Travis test still seems to be hanging when asking for admin privileges. Though now it seems to also be hanging on my local test as well.

**Comment by miccal** [comment]
> Though now it hangs on my local test as well without asking for a password.

So it was working for you locally before, without `sudo: true`?

**Comment by ruiokada** [comment]
> So it was working for you locally before, without `sudo: true`?

It was working in the sense that it would do an OS password prompt for administrator privileges and then proceed with the uninstall. I'm guessing it would be preferable to have it prompt for sudo privileges in terminal but this makes the script hang for some reason.

**Comment by ran-dall** [comment]
@miccal @ruiokada Where are we at with this?

**Comment by miccal** [comment]
@ran-dall tests are failing, but I am not sure why.

**Comment by ran-dall** [comment]
@miccal I think I've seen this before with another uninstaller script. The only solution I see is to run it without `sudo: true` because as it sits, it doesn't work. It didn't work for me locally, just now.

**Comment by ran-dall** [comment]
@ruiokada @miccal Now with this change (3eb9d34), I can successfully uninstall but yes, it does ask me for the password; but it beats hanging indefinitely.

**Comment by ruiokada** [comment]
@ran-dall Is it fine that it doesn't pass the CI tests?

**Comment by ran-dall** [comment]
> @ran-dall Is it fine that it doesn't pass the CI tests?

@ruiokada Absolutely, in these cases we just have to manually review and merge it. Regardless, the CI is really supposed to be more an assistant, as opposed to anything else. However, IMO it is worth the effort to try to make it happy when possible. 😉

**Pull request merged**